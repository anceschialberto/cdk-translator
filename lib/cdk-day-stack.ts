/* eslint-disable @typescript-eslint/no-empty-function */
import * as cdk from "@aws-cdk/core";
import { HttpApi, HttpMethod } from "@aws-cdk/aws-apigatewayv2";
import { LambdaProxyIntegration } from "@aws-cdk/aws-apigatewayv2-integrations";
import { AttributeType, Table } from "@aws-cdk/aws-dynamodb";
import { EventBus, Rule } from "@aws-cdk/aws-events";
import { LambdaFunction } from "@aws-cdk/aws-events-targets";
import { Policy, PolicyStatement } from "@aws-cdk/aws-iam";
import { Runtime } from "@aws-cdk/aws-lambda";
import { NodejsFunction } from "@aws-cdk/aws-lambda-nodejs";
import { CfnOutput } from "@aws-cdk/core";

export class CdkDayStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // ###################################################
    // Translation DDB table
    // ###################################################
    const translateTable = new Table(this, "TranslateTable", {
      partitionKey: { name: "id", type: AttributeType.STRING },
      sortKey: { name: "language", type: AttributeType.STRING },
    });

    // ###################################################
    // Translation EventBridge bus
    // ###################################################
    const translateBus = new EventBus(this, "TranslateBus", {
      eventBusName: "TranslateBus",
    });

    // ###################################################
    // Put translation function
    // ###################################################
    const putTranslationFunction = new NodejsFunction(
      this as any,
      "PutTranslationFunction",
      {
        runtime: Runtime.NODEJS_14_X,
        entry: "src/put-translation/app.ts",
        handler: "handler",
        depsLockFilePath: "src/put-translation/package-lock.json",
        bundling: {
          commandHooks: {
            beforeBundling(inputDir: string, outputDir: string): any {},
            afterBundling(inputDir: string, outputDir: string): string[] {
              return [`cp ${inputDir}/package.json ${outputDir}`];
            },
            beforeInstall(inputDir: string, outputDir: string): any {},
          },
        },
        environment: {
          TRANSLATE_BUS: translateBus.eventBusName,
        },
      }
    );

    translateBus.grantPutEventsTo(putTranslationFunction);

    const translatePolicyStatement = new PolicyStatement({
      actions: ["translate:TranslateText"],
      resources: ["*"],
    });

    putTranslationFunction.role?.attachInlinePolicy(
      new Policy(this, "PutTranslatePolicy", {
        statements: [translatePolicyStatement],
      })
    );

    // ###################################################
    // Get translations function
    // ###################################################
    const getTranslationFunction = new NodejsFunction(
      this as any,
      "GetTranslationFunction",
      {
        runtime: Runtime.NODEJS_14_X,
        entry: "src/get-translation/app.ts",
        handler: "handler",
        depsLockFilePath: "src/get-translation/package-lock.json",
        bundling: {
          commandHooks: {
            beforeBundling(inputDir: string, outputDir: string): any {},
            afterBundling(inputDir: string, outputDir: string): string[] {
              return [`cp ${inputDir}/package.json ${outputDir}`];
            },
            beforeInstall(inputDir: string, outputDir: string): any {},
          },
        },
        environment: {
          TRANSLATE_TABLE: translateTable.tableName,
        },
      }
    );

    translateTable.grantReadData(getTranslationFunction);

    // ###################################################
    // Save translations function
    // ###################################################
    const saveTranslationFunction = new NodejsFunction(
      this as any,
      "SaveTranslationFunction",
      {
        runtime: Runtime.NODEJS_14_X,
        entry: "src/save-translation/app.ts",
        handler: "handler",
        depsLockFilePath: "src/save-translation/package-lock.json",
        bundling: {
          commandHooks: {
            beforeBundling(inputDir: string, outputDir: string): any {},
            afterBundling(inputDir: string, outputDir: string): string[] {
              return [`cp ${inputDir}/package.json ${outputDir}`];
            },
            beforeInstall(inputDir: string, outputDir: string): any {},
          },
        },
        environment: {
          TRANSLATE_TABLE: translateTable.tableName,
        },
      }
    );

    translateTable.grantWriteData(saveTranslationFunction);

    // ###################################################
    // EventBridge Rule
    // ###################################################
    new Rule(this, "SaveTranslationRule", {
      eventBus: translateBus,
      eventPattern: { detailType: ["translation"] },
      targets: [new LambdaFunction(saveTranslationFunction)],
    });

    // ###################################################
    // API Gateway and routes
    // ###################################################
    const translateAPI = new HttpApi(this, "TranslateAPI");

    translateAPI.addRoutes({
      path: "/",
      methods: [HttpMethod.POST],
      integration: new LambdaProxyIntegration({
        handler: putTranslationFunction,
      }),
    });

    const getProxy = new LambdaProxyIntegration({
      handler: getTranslationFunction,
    });

    translateAPI.addRoutes({
      path: "/{id}",
      methods: [HttpMethod.GET],
      integration: getProxy,
    });

    translateAPI.addRoutes({
      path: "/",
      methods: [HttpMethod.GET],
      integration: getProxy,
    });

    // ###################################################
    // Outputs
    // ###################################################
    new CfnOutput(this, "API url", {
      value: translateAPI.url!,
    });
    new CfnOutput(this, "Put Function Name", {
      value: putTranslationFunction.functionName,
    });
    new CfnOutput(this, "Save Function Name", {
      value: saveTranslationFunction.functionName,
    });
    new CfnOutput(this, "Get Function Name", {
      value: getTranslationFunction.functionName,
    });
    new CfnOutput(this, "Translation Bus", {
      value: translateBus.eventBusName,
    });
    new CfnOutput(this, "Translation Table", {
      value: translateTable.tableName,
    });
  }
}
