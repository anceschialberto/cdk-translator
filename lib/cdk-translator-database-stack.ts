/* eslint-disable @typescript-eslint/no-empty-function */
import { Construct } from "constructs";
import { CfnOutput, Stack, StackProps } from "aws-cdk-lib"; // core constructs
import { aws_dynamodb as dynamodb } from "aws-cdk-lib";

export class CdkTranslatorDatabaseStack extends Stack {
  readonly translateTable: dynamodb.Table;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // ###################################################
    // Translation DDB table
    // ###################################################
    this.translateTable = new dynamodb.Table(this, "TranslateTable", {
      partitionKey: { name: "id", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "language", type: dynamodb.AttributeType.STRING },
    });

    // ###################################################
    // Outputs
    // ###################################################
    new CfnOutput(this, "Translation Table", {
      value: this.translateTable.tableName,
    });
  }
}
