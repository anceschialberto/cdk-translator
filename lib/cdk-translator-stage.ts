import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { CdkTranslatorDatabaseStack } from "./cdk-translator-database-stack";
import { CdkTranslatorStack } from "./cdk-translator-stack";

/**
 * The application
 *
 * May consist of one or more Stacks (here, only one)
 */
export class CdkTranslatorStage extends cdk.Stage {
  constructor(scope: Construct, id: string, props?: cdk.StageProps) {
    super(scope, id, props);

    const database = new CdkTranslatorDatabaseStack(
      this,
      "CdkTranslatorDatabaseStack",
      {
        terminationProtection: true,
      }
    );

    new CdkTranslatorStack(this, "CdkTranslatorStack", {
      translateTable: database.translateTable,
    });
  }
}
