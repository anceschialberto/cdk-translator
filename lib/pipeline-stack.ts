import { Construct } from "constructs";
import { Environment, Stack, StackProps } from "aws-cdk-lib"; // core constructs

import { aws_codestarconnections as codestarconnections } from "aws-cdk-lib";
import {
  CodePipeline,
  CodePipelineSource,
  ManualApprovalStep,
  ShellStep,
} from "aws-cdk-lib/pipelines";

import { CdkTranslatorStage } from "./cdk-translator-stage";

interface PipelineStackProps extends StackProps {
  devEnvironment: Environment;
  prodEnvironment: Environment;
}

export class PipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: PipelineStackProps) {
    super(scope, id, props);

    const githubConnection = new codestarconnections.CfnConnection(
      this,
      "GitHubConnection",
      {
        connectionName: "github-connection",
        providerType: "GitHub",
      }
    );

    const source = CodePipelineSource.connection(
      "anceschialberto/cdk-translator",
      "main",
      { connectionArn: githubConnection.attrConnectionArn }
    );

    const synthStep = new ShellStep("Synth", {
      input: source,
      commands: ["npm i -g npm@8", "npm ci", "npx cdk synth"],
    });

    // Pipeline creation starts
    const pipeline = new CodePipeline(this, "Pipeline", {
      synth: synthStep,
      crossAccountKeys: true,
    });

    const devStage = new CdkTranslatorStage(this, "Dev", {
      env: props?.devEnvironment,
    });

    pipeline.addStage(devStage, {
      post: [
        new ManualApprovalStep("DeployApproval", {
          comment: "Approve to deploy to the next stage",
        }),
      ],
    });

    const prodStage = new CdkTranslatorStage(this, "Prod", {
      env: props?.prodEnvironment,
    });

    pipeline.addStage(prodStage);
  }
}
