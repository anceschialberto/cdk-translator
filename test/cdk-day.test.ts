/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-call */

import { Template } from "@aws-cdk/assertions";
import * as cdk from "@aws-cdk/core";
import * as CdkDay from "../lib/cdk-day-stack";

test("Empty Stack", () => {
  const app = new cdk.App();
  const stack = new CdkDay.CdkDayStack(app, "CdkDayStack");

  // Prepare the stack for assertions.
  const template = Template.fromStack(stack);

  // Assert the template matches the snapshot.
  expect(template.toJSON()).toMatchSnapshot();
});
