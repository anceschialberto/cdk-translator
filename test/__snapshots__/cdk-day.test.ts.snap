// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty Stack 1`] = `
Object {
  "Outputs": Object {
    "APIurl": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "TranslateAPI19A43816",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
          ],
        ],
      },
    },
    "GetFunctionName": Object {
      "Value": Object {
        "Ref": "GetTranslationFunction0677F2E3",
      },
    },
    "PutFunctionName": Object {
      "Value": Object {
        "Ref": "PutTranslationFunction9E955411",
      },
    },
    "SaveFunctionName": Object {
      "Value": Object {
        "Ref": "SaveTranslationFunctionD9E440D6",
      },
    },
    "TranslationBus": Object {
      "Value": Object {
        "Ref": "TranslateBus46D6DDAA",
      },
    },
    "TranslationTable": Object {
      "Value": Object {
        "Ref": "TranslateTable1ABF9811",
      },
    },
  },
  "Parameters": Object {
    "AssetParameters524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091eArtifactHash0434C292": Object {
      "Description": "Artifact hash for asset \\"524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091e\\"",
      "Type": "String",
    },
    "AssetParameters524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091eS3Bucket89A88BC5": Object {
      "Description": "S3 bucket for asset \\"524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091e\\"",
      "Type": "String",
    },
    "AssetParameters524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091eS3VersionKey8E2F5365": Object {
      "Description": "S3 key for asset version \\"524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091e\\"",
      "Type": "String",
    },
    "AssetParameters5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0ArtifactHash4F1FB23D": Object {
      "Description": "Artifact hash for asset \\"5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0\\"",
      "Type": "String",
    },
    "AssetParameters5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0S3Bucket5C7AE909": Object {
      "Description": "S3 bucket for asset \\"5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0\\"",
      "Type": "String",
    },
    "AssetParameters5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0S3VersionKeyB77D2116": Object {
      "Description": "S3 key for asset version \\"5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0\\"",
      "Type": "String",
    },
    "AssetParameters5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8eArtifactHash50290F32": Object {
      "Description": "Artifact hash for asset \\"5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8e\\"",
      "Type": "String",
    },
    "AssetParameters5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8eS3Bucket438FB6D1": Object {
      "Description": "S3 bucket for asset \\"5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8e\\"",
      "Type": "String",
    },
    "AssetParameters5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8eS3VersionKeyB6076E07": Object {
      "Description": "S3 key for asset version \\"5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8e\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GetTranslationFunction0677F2E3": Object {
      "DependsOn": Array [
        "GetTranslationFunctionServiceRoleDefaultPolicy7862AC6E",
        "GetTranslationFunctionServiceRole408CD69C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8eS3Bucket438FB6D1",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8eS3VersionKeyB6076E07",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5e3aa02ee8341cc59b83b76bbfe65f66e86dc6d8592ec9d838f5f0527de63a8eS3VersionKeyB6076E07",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TRANSLATE_TABLE": Object {
              "Ref": "TranslateTable1ABF9811",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GetTranslationFunctionServiceRole408CD69C",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "GetTranslationFunctionServiceRole408CD69C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GetTranslationFunctionServiceRoleDefaultPolicy7862AC6E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TranslateTable1ABF9811",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetTranslationFunctionServiceRoleDefaultPolicy7862AC6E",
        "Roles": Array [
          Object {
            "Ref": "GetTranslationFunctionServiceRole408CD69C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PutTranslatePolicyB54B2023": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "translate:TranslateText",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PutTranslatePolicyB54B2023",
        "Roles": Array [
          Object {
            "Ref": "PutTranslationFunctionServiceRoleE038FB93",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PutTranslationFunction9E955411": Object {
      "DependsOn": Array [
        "PutTranslationFunctionServiceRoleDefaultPolicyB878D483",
        "PutTranslationFunctionServiceRoleE038FB93",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091eS3Bucket89A88BC5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091eS3VersionKey8E2F5365",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters524794beb8ee84510e71cd10847fefe4503a5c8277a2d0bd4eef23761df7091eS3VersionKey8E2F5365",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TRANSLATE_BUS": Object {
              "Ref": "TranslateBus46D6DDAA",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "PutTranslationFunctionServiceRoleE038FB93",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "PutTranslationFunctionServiceRoleDefaultPolicyB878D483": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "events:PutEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TranslateBus46D6DDAA",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PutTranslationFunctionServiceRoleDefaultPolicyB878D483",
        "Roles": Array [
          Object {
            "Ref": "PutTranslationFunctionServiceRoleE038FB93",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PutTranslationFunctionServiceRoleE038FB93": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SaveTranslationFunctionAllowEventRuleCdkDayStackSaveTranslationRule8F180B16B8FEB8E6": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "SaveTranslationFunctionD9E440D6",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "SaveTranslationRule77BFC15F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SaveTranslationFunctionD9E440D6": Object {
      "DependsOn": Array [
        "SaveTranslationFunctionServiceRoleDefaultPolicyE1167723",
        "SaveTranslationFunctionServiceRoleDCF57C08",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0S3Bucket5C7AE909",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0S3VersionKeyB77D2116",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters5872e356d8dbde47c8970f0c102f0fd792939010e9514d555fb04587237117a0S3VersionKeyB77D2116",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TRANSLATE_TABLE": Object {
              "Ref": "TranslateTable1ABF9811",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "SaveTranslationFunctionServiceRoleDCF57C08",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "SaveTranslationFunctionServiceRoleDCF57C08": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SaveTranslationFunctionServiceRoleDefaultPolicyE1167723": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TranslateTable1ABF9811",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SaveTranslationFunctionServiceRoleDefaultPolicyE1167723",
        "Roles": Array [
          Object {
            "Ref": "SaveTranslationFunctionServiceRoleDCF57C08",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SaveTranslationRule77BFC15F": Object {
      "Properties": Object {
        "EventBusName": Object {
          "Ref": "TranslateBus46D6DDAA",
        },
        "EventPattern": Object {
          "detail-type": Array [
            "translation",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "SaveTranslationFunctionD9E440D6",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "TranslateAPI19A43816": Object {
      "Properties": Object {
        "Name": "TranslateAPI",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "TranslateAPIDefaultStage3BDD98F9": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TranslateAPI19A43816",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "TranslateAPIGET71FB00DE": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TranslateAPI19A43816",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "GET /",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "TranslateAPIGETidHttpIntegration880cc39dcb422e77f6801a8ce2294b073770B841",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "TranslateAPIGETCdkDayStackTranslateAPIGETCB27EC2BPermission8478C60C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GetTranslationFunction0677F2E3",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TranslateAPI19A43816",
              },
              "/*/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TranslateAPIGETid45762B1E": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TranslateAPI19A43816",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "GET /{id}",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "TranslateAPIGETidHttpIntegration880cc39dcb422e77f6801a8ce2294b073770B841",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "TranslateAPIGETidCdkDayStackTranslateAPIGETid7794A409Permission2E72401E": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GetTranslationFunction0677F2E3",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TranslateAPI19A43816",
              },
              "/*/*/{id}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TranslateAPIGETidHttpIntegration880cc39dcb422e77f6801a8ce2294b073770B841": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TranslateAPI19A43816",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "GetTranslationFunction0677F2E3",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "TranslateAPIPOSTCdkDayStackTranslateAPIPOST457D5C02Permission726A551D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "PutTranslationFunction9E955411",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TranslateAPI19A43816",
              },
              "/*/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TranslateAPIPOSTD32A2B26": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TranslateAPI19A43816",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "TranslateAPIPOSTHttpIntegration8710e6b71ff5c67e39e9b5b40fe39dbaEFCE3F57",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "TranslateAPIPOSTHttpIntegration8710e6b71ff5c67e39e9b5b40fe39dbaEFCE3F57": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TranslateAPI19A43816",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "PutTranslationFunction9E955411",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "TranslateBus46D6DDAA": Object {
      "Properties": Object {
        "Name": "TranslateBus",
      },
      "Type": "AWS::Events::EventBus",
    },
    "TranslateTable1ABF9811": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "language",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "language",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
